prow/pod-utils/gcs/upload.go

```diff
@@ -215,29 +241,39 @@ type dataWriter interface {
 	io.WriteCloser
 	fullUploadPath() string
 	ApplyWriterOptions(opts pkgio.WriterOptions)
+	compressData() bool
 }
 
 type openerObjectWriter struct {
 	pkgio.Opener
-	Context     context.Context
-	Bucket      string
-	Dest        string
-	opts        []pkgio.WriterOptions
-	writeCloser pkgio.WriteCloser
+	Context       context.Context
+	Bucket        string
+	Dest          string
+	compress      bool
+	opts          []pkgio.WriterOptions
+	storageWriter pkgio.WriteCloser
+	zipWriter     pkgio.WriteCloser
 }
 
 func (w *openerObjectWriter) Write(p []byte) (n int, err error) {
-	if w.writeCloser == nil {
-		w.writeCloser, err = w.Opener.Writer(w.Context, w.fullUploadPath(), w.opts...)
+	if w.storageWriter == nil {
+		w.storageWriter, err = w.Opener.Writer(w.Context, w.fullUploadPath(), w.opts...)
 		if err != nil {
 			return 0, err
 		}
+		if w.compress {
+			w.zipWriter = gzip.NewWriter(w.storageWriter)
```

Comment:
Why should we prefer to have both and choose each at every call-site as opposed to just having one `pkgio.WriteCloser` at the top-level when creating a `openerObjectWriter` ?

Reply:
I also assumed that would be the way to do it, but I found in practice that both writers must be closed. The wrapping writer doesn't seem to close the inner one when it itself is closed. This caused the files to not actually be uploaded.

Reply:
Hm, perhaps then we can store one writer and one closer? If `w.zipWriter` is present, then it's always invalid to write to `w.storageWriter` and vice versa. Holding on to two `io.Writer` seems like a weird state to be in.

Reply:
I'll play around with it and see if I can figure something better out. I think that storing one writer and one closer won't help fix the original problem though. Close will still need to be called on both, which will be even more complex.

Reply:
I verified that this is [documented behavior](https://pkg.go.dev/compress/gzip#Writer.Close). I can't think of any way to do this without storing references to both writers in `openerObjectWriter`. The references (and the type itself) aren't exported so the usages are contained to `Write` and `Close`.

Reply:
@smg247 do we ever need to *write* to the object handle? Sorry - I did not mean to have just one `io.Closer` - but two. It's just that if we're in compression mode it will always be incorrect to write to the other writer, right?

Reply:
Ah, I understand what you mean now. That will work. I just took care of that update.